In [1]:
import sys
import os
sys.path.append(os.path.abspath('../scripts'))

# Get the current directory
current_dir = os.getcwd()
print(current_dir)

# Get the parent directory
parent_dir = os.path.dirname(current_dir)
print(parent_dir)

# Insert the path to the parent directory
sys.path.insert(0, parent_dir)

from load_data import load_data_using_sqlalchemy

c:\Users\HP\Documents\10-Academy\Telecom-WEEK2\notebooks
c:\Users\HP\Documents\10-Academy\Telecom-WEEK2


database connection and loading data

In [2]:
# Define your SQL query
query = "SELECT * FROM xdr_data;"  # Replace with your actual table name

# Load data from PostgreSQL using SQLAlchemy
df = load_data_using_sqlalchemy(query)

# Display the first few rows of the dataframe
if df is not None:
    print("Successfully loaded the data")
else:
    print("Failed to load data.")

Successfully loaded the data


In [3]:
from scripts.user_overview_analysis import OverviewAnalyzer


In [4]:
analyzer = OverviewAnalyzer(df, 'HTTP DL (Bytes)')

In [5]:
missing_values_df = analyzer.missing_values_table(df)
print("Missing Values in df:")
print(missing_values_df)

Your selected dataframe has 55 columns.
There are 41 columns that have missing values.
Missing Values in df:
                                          Missing Values  % of Total Values  \
Nb of sec with 37500B < Vol UL                    130254               86.8   
Nb of sec with 6250B < Vol UL < 37500B            111843               74.6   
Nb of sec with 125000B < Vol DL                    97538               65.0   
TCP UL Retrans. Vol (Bytes)                        96649               64.4   
Nb of sec with 31250B < Vol DL < 125000B           93586               62.4   
Nb of sec with 1250B < Vol UL < 6250B              92894               61.9   
Nb of sec with 6250B < Vol DL < 31250B             88317               58.9   
TCP DL Retrans. Vol (Bytes)                        88146               58.8   
HTTP UL (Bytes)                                    81810               54.5   
HTTP DL (Bytes)                                    81474               54.3   
Avg RTT DL (ms)       

Analysis of Missing Values in the Telecom Dataset
The dataset contains 55 columns, of which 41 columns have missing values. This high proportion of missing data across numerous features is a significant concern, as it can impact the quality and reliability of the analysis. Here's a breakdown of the findings:

Key Findings:
Significant Missing Data in High-Volume Features:

Columns like Nb of sec with 37500B < Vol UL, Nb of sec with 6250B < Vol UL < 37500B, and Nb of sec with 125000B < Vol DL have substantial missing values, with percentages of missing data as high as 86.8%, 74.6%, and 65.0%, respectively.
These columns are critical as they represent key performance indicators related to data upload and download volumes. The high percentage of missing data suggests either data collection issues or irregularities in user activity.
Transmission and HTTP Data:

Columns related to TCP retransmission and HTTP data, such as TCP UL Retrans. Vol (Bytes) and HTTP UL (Bytes), also exhibit a high percentage of missing values (around 54-64%).
Missing data in these features might hinder the understanding of network performance and user behavior, especially in analyzing retransmission rates and HTTP traffic.
Moderate Missing Data in Handset Information:

The Handset Manufacturer, Handset Type, and IMEI columns have around 0.4% missing data. Although the percentage is relatively low, it is crucial to address these missing values as they are fundamental to the User Overview analysis, which focuses on handset types and manufacturers.
Accurate and complete handset information is essential for identifying trends in device usage and making informed marketing recommendations.
Critical Identifiers and Time Features:

Columns such as MSISDN/Number, Bearer Id, Start, End, and Dur. (ms) show minimal missing values (less than 1%). These columns are vital for user identification, session tracking, and duration analysis.
Ensuring these features are complete and accurate is critical for analyzing user behavior and session durations.
Other Technical Metrics:

Several columns, such as UL TP < 10 Kbps (%), DL TP > 1 Mbps (%), and Total UL (Bytes), have very minimal missing values (less than 1%).
These columns provide detailed insights into data transmission rates and overall data usage, which are crucial for performance analysis.
Interpretation:
The dataset has significant missing values in key performance and volume-related features, which could potentially skew the analysis if not properly addressed. While some features have minimal missing data, the overall high percentage of missing data across the dataset indicates potential data collection issues, user behavior irregularities, or device-related problems.

Recommendations:
Data Imputation or Exclusion: Consider imputation techniques for missing values where appropriate, especially for columns with a smaller percentage of missing data. For features with extremely high missing values (over 50%), assess the impact on the analysis and consider excluding these columns if they do not add substantial value.

Focus on Handset Data: Ensure that the handset information is as complete as possible since it is crucial for the User Overview analysis. Missing data in these columns should be addressed with targeted imputation strategies or by investigating the source of the missing data.

Investigate Data Collection: The high level of missing data in certain columns may indicate issues with the data collection process. Investigating the root causes of these missing values could help improve data quality in future collections.

Sensitivity Analysis: Conduct sensitivity analysis to understand how the missing data might impact the results. This can help in determining the robustness of your findings and ensuring that the analysis remains reliable even with incomplete data.

This analysis provides an overview of the dataset’s condition and highlights areas where improvements or careful consideration are necessary for a more accurate and insightful analysis.

The telecom dataset I'm working with contains 55 columns, and I've found that 41 of these columns have missing values. This is concerning, especially since key features related to data upload and download volumes, like Nb of sec with 37500B < Vol UL and Nb of sec with 125000B < Vol DL, have missing values ranging from 65% to nearly 87%. The high percentage of missing data in these areas could indicate problems with data collection or irregularities in user activity. I've also noticed that important columns related to TCP retransmission and HTTP data, such as TCP UL Retrans. Vol (Bytes) and HTTP UL (Bytes), have missing values of around 54-64%. This could affect my analysis of network performance and user behavior. On a more positive note, the handset-related data, such as Handset Manufacturer and Handset Type, has a much lower percentage of missing values, around 0.4%. However, even this small amount of missing data needs to be addressed since it's crucial for my analysis of device usage. Columns that are vital for tracking user sessions and behavior, like MSISDN/Number and Bearer Id, have minimal missing values, which is reassuring. Given these findings, I'm considering using imputation techniques for columns with lower percentages of missing data, while possibly excluding those with extremely high missing values. I also think it's important to investigate the causes of these missing values to improve data quality in future projects. Overall, I'm aware that these missing values could impact my analysis, so I'll be conducting sensitivity checks to ensure that my findings are as accurate and reliable as possible.

In [6]:
df['x'] = analyzer.convert_bytes_to_megabytes(df, 'HTTP DL (Bytes)')

print(df)

           Bearer Id            Start  Start ms              End  End ms  \
0       1.311448e+19   4/4/2019 12:01     770.0  4/25/2019 14:35   662.0   
1       1.311448e+19   4/9/2019 13:04     235.0   4/25/2019 8:15   606.0   
2       1.311448e+19   4/9/2019 17:42       1.0  4/25/2019 11:58   652.0   
3       1.311448e+19   4/10/2019 0:31     486.0   4/25/2019 7:36   171.0   
4       1.311448e+19  4/12/2019 20:10     565.0  4/25/2019 10:40   954.0   
...              ...              ...       ...              ...     ...   
149996  7.277826e+18   4/29/2019 7:28     451.0   4/30/2019 6:02   214.0   
149997  7.349883e+18   4/29/2019 7:28     483.0  4/30/2019 10:41   187.0   
149998  1.311448e+19   4/29/2019 7:28     283.0  4/30/2019 10:46   810.0   
149999  1.311448e+19   4/29/2019 7:28     696.0  4/30/2019 10:40   327.0   
150000           NaN             None       NaN             None     NaN   

        Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0       1823652.0  2.08

In [8]:
user_behavior = analyzer.aggregate_user_behavior(df)
print(user_behavior.head())
print(df)

   MSISDN/Number  num_xdr_sessions  total_session_duration  \
0   3.360100e+10                 1                116720.0   
1   3.360100e+10                 1                181230.0   
2   3.360100e+10                 1                134969.0   
3   3.360101e+10                 1                 49878.0   
4   3.360101e+10                 2                 37104.0   

   total_download_data  total_upload_data  social_media_dl  social_media_ul  \
0         8.426375e+08         36053108.0        2206504.0          25631.0   
1         1.207552e+08         36104459.0        2598548.0          62017.0   
2         5.566597e+08         39306820.0        3148004.0          47619.0   
3         4.019932e+08         20327526.0         251469.0          28825.0   
4         1.363130e+09         94280527.0        2861230.0          51312.0   

   youtube_dl  youtube_ul  netflix_dl  ...   google_dl  google_ul   email_dl  \
0  14900201.0   6724347.0  10265105.0  ...   3337123.0  1051882.0   8374

Based on the aggregated results provided, here's a summary of user behavior on the applications:

Number of xDR Sessions: Each user (identified by their MSISDN/Number) has a different number of xDR sessions recorded, which reflects how many distinct sessions they participated in. For instance, the first few users in the dataset each have only one recorded session, while one user has two.

Session Duration: The total duration of these sessions varies significantly, from around 37,104 milliseconds to over 1,340,000 milliseconds. This indicates varying levels of engagement or activity across different users.

Total Data Volume (Download and Upload): The total data volume (both download and upload) for these sessions also varies. Some users have downloaded and uploaded hundreds of megabytes to several gigabytes of data during their sessions. For example, the total download data for one user is as high as 1.36 GB, while another user downloaded around 402 MB.

Application-Specific Data Usage: The data shows how much data was consumed by various applications (Social Media, YouTube, Netflix, Google, Email, Gaming, and Others) during these sessions.

Social Media Data: Users have used between 2206504 and 2861230 bytes for downloading and uploading social media content.
YouTube Data: YouTube data usage shows substantial variance, with download volumes reaching up to around 10 MB for some users.
Netflix Data: Netflix also shows considerable usage, with data downloads going up to around 24.9 MB for one user.
Google, Email, Gaming, and Other Data: These categories similarly exhibit significant data usage, with Google data usage being particularly high in some cases.
Total Application Data Volume: The total application data volume column sums up the data usage across all applications for each user. The total application data volume ranges from around 436 MB to over 2.23 GB. This indicates how much data each user consumed across all tracked applications during their xDR sessions.

Conclusion
This summary provides insights into user behavior, showing that some users are significantly more active in terms of data usage and session duration, particularly in applications like Social Media, YouTube, and Netflix. This information can be valuable for understanding user engagement patterns and identifying which applications are driving the most data consumption.